### PR TITLE
fix: Zero sum fix

### PR DESCRIPTION
**Pull request by lazovicff**
## Changes
- In case of no outbound local trust, distribute to seed peers
- in case of no seed peers, distribute seed trust to everyone equally
- Optimisations to the core algo

## Reminders
- [x] Add a prefix to PR: `feat:` (for features), `chore:` (for chores/refactoring), `fix:` (for bug fixes), or any from: https://www.conventionalcommits.org/en/v1.0.0/#summary
- [x] Assign reviewers
- [x] Link related issues
- [x] Make it draft if it is still WIP
- [x] Write unit tests for your code if possible
- [x] Do manual testing and make sure your changes don't break the rest of the codebase

### CODE REVIEW

File: common/src/algos/et.rs [hunk 1]
```diff
@@ -23,14 +37,31 @@ fn pre_process(lt: &mut HashMap<(u64, u64), f32>) {
...
     }
 
+    for from in all_peers.iter() {
+        let sum = sum_map.get(from).unwrap_or(&0.0);
+        // If peer does not have outbound trust,
+        // his trust will be distributed to seed peers based on their seed/pre-trust
+        if *sum == 0.0 {
+            for (to, value) in seed {
+                lt.insert((*from, *to), *value);
+            }
```
Comment (duguorong009) [inline common/src/algos/et.rs:60]: I have a question.
If `from` == `to`, it means giving trust value to itself.
Then, we should call `pre_process` after `normalize_lt`. 
Is it right?

File: common/src/algos/et.rs [hunk 2]
```diff
@@ -39,33 +70,53 @@ fn normalise_lt(lt: &mut HashMap<(u64, u64), f32>) -> Result<(), algos::Error> {
...
+) -> Result<(), algos::Error> {
     // Calculate the sum of all seed trust values.
     let sum: f32 = seed.iter().map(|(_, v)| v).sum();
 
     // Divide each seed trust value by the sum to normalise.
     if sum == 0.0 {
-        return Err(algos::Error::ZeroSum);
+        for i in all_peers {
+            seed.insert(*i, 1.0);
+        }
```
Comment (duguorong009) [inline common/src/algos/et.rs:88]: ditto.

File: common/src/algos/et.rs [hunk 3]
```diff
@@ -39,33 +70,53 @@ fn normalise_lt(lt: &mut HashMap<(u64, u64), f32>) -> Result<(), algos::Error> {
...
-pub fn positive_run<const NUM_ITER: usize>(
+pub fn positive_run(
     mut lt: HashMap<(u64, u64), f32>, mut seed: HashMap<u64, f32>,
 ) -> Result<Vec<(u64, f32)>, algos::Error> {
+    let all_peers = get_all_peers(&lt, &seed);
     pre_process(&mut lt);
-    normalise_lt(&mut lt)?;
-    normalise_seed(&mut seed)?;
+    normalise_seed(&all_peers, &mut seed)?;
+    normalise_lt(&all_peers, &mut lt, &seed)?;
```
Comment (duguorong009) [inline common/src/algos/et.rs:119]: ditto.

File: common/src/algos/et.rs [hunk 4]
```diff
@@ -132,10 +183,12 @@ pub fn is_converged_org(scores: &HashMap<String, f32>, next_scores: &HashMap<Str
...
 pub fn convergence_check(
-    mut lt: HashMap<(u64, u64), f32>, seed: &HashMap<u64, f32>, scores: &HashMap<u64, f32>,
+    mut lt: HashMap<(u64, u64), f32>, mut seed: HashMap<u64, f32>, scores: &HashMap<u64, f32>,
 ) -> Result<bool, algos::Error> {
-    // Normalize the local trust matrix
-    normalise_lt(&mut lt)?;
+    let all_peers = get_all_peers(&lt, &seed);
+    pre_process(&mut lt);
+    normalise_seed(&all_peers, &mut seed)?;
+    normalise_lt(&all_peers, &mut lt, &seed)?;
```
Comment (duguorong009) [inline common/src/algos/et.rs:191]: ditto.

File: verifier/src/runner.rs [hunk 5]
```diff
@@ -353,7 +353,7 @@ impl VerificationRunner {
             }
             score_entries_map
         };
-        convergence_check(lt.clone(), seed, &score_entries).map_err(Error::Algo)
+        convergence_check(lt.clone(), seed.clone(), &score_entries).map_err(Error::Algo)
```
Comment (astralblue) [inline verifier/src/runner.rs:356]: Nice catch, would've jeopardized seed update after initial use.

**Review by astralblue — APPROVED** [review]
LGTM

File: verifier/src/runner.rs (hunk 5 above)
Comment (lazovicff) [reply verifier/src/runner.rs:356]: No, we didn't, we just passed the reference, but now we pass the value by cloning it

**Pull request merged**